Enhancement proposal: expansion of ${USER} variable in 'authfile' module argument

This may be related to #152 ,

As the title suggests, it would be nice to have the ${USER} variable expansion in the `authfile` module argument to allow for multiple users having their own sub-directory with a common structure; this would be paired with the `openasuser` module argument like this:

```
auth sufficient pam_u2f.so authfile=/etc/2fa/u2f/${USER}/u2f_mappings openasuser
```

This would allow for this file structure:

```
/etc/2fa/
└── u2f/
    └── ${USER}/
        └── u2f_mappings
```

Each `u2f_mappings` would need their according `0400` access permissions, as well as the correct user and group for each user's mappings.

This would allow for the benefits of a centralized and an individual authorization file (usage of an encrypted home directory, for instance).

I have tested both with and without the `openasuser` option (having the mappings associated to the user, or to `root`).  Both worked like a charm!

The only observation that I can think of making is to perhaps make a small note to mention that, in some cases, `libpam0g-dev` provides `libpam-dev`.

Here is some of the `stderr` output in case it is useful:

<details>
<summary><code>autoreconf --install</code></summary>

```console
user@computer:~/projects/pam-u2f$ autoreconf --install 
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: copying file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:17: installing 'build-aux/ar-lib'
configure.ac:15: installing 'build-aux/compile'
configure.ac:8: installing 'build-aux/config.guess'
configure.ac:8: installing 'build-aux/config.sub'
configure.ac:10: installing 'build-aux/install-sh'
configure.ac:10: installing 'build-aux/missing'
Makefile.am: installing 'build-aux/depcomp'
parallel-tests: installing 'build-aux/test-driver'
```

</details>


<details>
<summary><code>./configure</code></summary>

```console
user@computer:~/projects/pam-u2f$ ./configure 
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for ar... ar
checking the archiver (ar) interface... ar
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for file... file
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking for security/pam_appl.h... yes
checking for security/pam_modules.h... yes
checking for security/pam_modutil.h... yes
checking for security/openpam.h... no
checking for pam_start in -lpam... yes
checking for pam_modutil_drop_priv... yes
checking for openpam_borrow_cred... no
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for libcrypto... yes
checking for libfido2 >= 1.3.0... yes
checking for secure_getenv... yes
checking for strlcpy... no
checking for readpassphrase... no
checking for explicit_bzero... yes
checking for memset_s... no
checking whether C compiler accepts -Werror=unknown-warning-option... no
checking whether C compiler accepts -Wall... yes
checking whether C compiler accepts -Wextra... yes
checking whether C compiler accepts -Wconversion... yes
checking whether C compiler accepts -Wconversion... (cached) yes
checking whether C compiler accepts -Wpedantic... yes
checking whether C compiler accepts -Wformat=2... yes
checking whether C compiler accepts -Wstrict-prototypes... yes
checking whether C compiler accepts -Wmissing-declarations... yes
checking whether C compiler accepts -Wmissing-prototypes... yes
checking whether C compiler accepts -Wnull-dereference... yes
checking whether C compiler accepts -Wshadow... yes
checking whether C compiler accepts -Wpointer-arith... yes
checking whether C compiler accepts -Wno-unused-but-set-variable... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating pamu2fcfg/Makefile
config.status: creating tests/Makefile
config.status: creating fuzz/Makefile
config.status: creating man/Makefile
config.status: creating tests/credentials/new_double_-N.cred
config.status: creating tests/credentials/new_double_-P-N.cred
config.status: creating tests/credentials/new_double_-P.cred
config.status: creating tests/credentials/new_double_-P-V-N.cred
config.status: creating tests/credentials/new_double_-P-V.cred
config.status: creating tests/credentials/new_double_-r-N.cred
config.status: creating tests/credentials/new_double_-r-P-N.cred
config.status: creating tests/credentials/new_double_-r-P.cred
config.status: creating tests/credentials/new_double_-r-P-V-N.cred
config.status: creating tests/credentials/new_double_-r-P-V.cred
config.status: creating tests/credentials/new_double_-r.cred
config.status: creating tests/credentials/new_double_-r-V-N.cred
config.status: creating tests/credentials/new_double_-r-V.cred
config.status: creating tests/credentials/new_double_.cred
config.status: creating tests/credentials/new_double_-V-N.cred
config.status: creating tests/credentials/new_double_-V.cred
config.status: creating tests/credentials/new_mixed_12.cred
config.status: creating tests/credentials/new_mixed_1-P2.cred
config.status: creating tests/credentials/new_mixed_-P12.cred
config.status: creating tests/credentials/new_mixed_-P1-P2.cred
config.status: creating tests/credentials/new_-N.cred
config.status: creating tests/credentials/new_-P-N.cred
config.status: creating tests/credentials/new_-P.cred
config.status: creating tests/credentials/new_-P-V-N.cred
config.status: creating tests/credentials/new_-P-V.cred
config.status: creating tests/credentials/new_-r-N.cred
config.status: creating tests/credentials/new_-r-P-N.cred
config.status: creating tests/credentials/new_-r-P.cred
config.status: creating tests/credentials/new_-r-P-V-N.cred
config.status: creating tests/credentials/new_-r-P-V.cred
config.status: creating tests/credentials/new_-r.cred
config.status: creating tests/credentials/new_-r-V-N.cred
config.status: creating tests/credentials/new_-r-V.cred
config.status: creating tests/credentials/new_.cred
config.status: creating tests/credentials/new_-V-N.cred
config.status: creating tests/credentials/new_-V.cred
config.status: creating tests/credentials/old_credential.cred
config.status: creating tests/credentials/ssh_credential.cred
config.status: creating tests/credentials/new_limited_count.cred
config.status: executing depfiles commands
config.status: executing libtool commands
configure: Summary of build options:

  Version:             1.2.2
  Host type:           x86_64-pc-linux-gnu
  Install prefix:      /usr/local
  Compiler:            gcc
  CFLAGS:              -g -O2
  CWFLAGS:             -Wall -Wextra -Wconversion -Wno-sign-conversion -Wpedantic -Wformat=2 -Wstrict-prototypes -Wmissing-declarations -Wmissing-prototypes -Wnull-dereference -Wshadow -Wpointer-arith
  CSFLAGS:             -Wno-unused-but-set-variable
  Library types:       Shared=yes, Static=no
  LIBFIDO2 CFLAGS:     
  LIBFIDO2 LIBS:       -lfido2 -lcrypto
  LIBCRYPTO CFLAGS:    
  LIBCRYPTO LIBS:      -lcrypto
  PAMDIR:              /lib/x86_64-linux-gnu/security

```

</details>


<details>
<summary><code>make</code> (warnings; Guix's python couldn't find asciidoc)</summary>

```console
user@computer:~/projects/pam-u2f$ make
Making all in .
make[1]: Entering directory '/home/user/projects/pam-u2f'
  CC       pam-u2f.lo
  CC       b64.lo
  CC       debug.lo
  CC       expand.lo
  CC       explicit_bzero.lo
  CC       util.lo
  CCLD     libmodule.la
  CCLD     pam_u2f.la
make[1]: Leaving directory '/home/user/projects/pam-u2f'
Making all in pamu2fcfg
make[1]: Entering directory '/home/user/projects/pam-u2f/pamu2fcfg'
  CC       pamu2fcfg.o
  CC       readpassphrase.o
readpassphrase.c: In function ‘readpassphrase’:
readpassphrase.c:133:23: warning: ignoring return value of ‘write’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  133 |                 (void)write(output, prompt, strlen(prompt));
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
readpassphrase.c:152:23: warning: ignoring return value of ‘write’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  152 |                 (void)write(output, "\n", 1);
      |                       ^~~~~~~~~~~~~~~~~~~~~~
  CC       strlcpy.o
  CC       ../util.o
  CC       ../b64.o
  CC       ../explicit_bzero.o
  CCLD     pamu2fcfg
make[1]: Leaving directory '/home/user/projects/pam-u2f/pamu2fcfg'
Making all in tests
make[1]: Entering directory '/home/user/projects/pam-u2f/tests'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/user/projects/pam-u2f/tests'
Making all in man
make[1]: Entering directory '/home/user/projects/pam-u2f/man'
  GEN      pamu2fcfg.1
/home/user/.guix-profile/bin/python3: Error while finding module specification for 'asciidoc.a2x' (ModuleNotFoundError: No module named 'asciidoc')
make[1]: *** [Makefile:581: pamu2fcfg.1] Error 1
make[1]: Leaving directory '/home/user/projects/pam-u2f/man'
make: *** [Makefile:782: all-recursive] Error 1
```

</details>


<details>
<summary><code>make</code> (fixed <code>PATH</code>)</summary>

```console
user@computer:~/projects/pam-u2f$ make
Making all in .
make[1]: Entering directory '/home/user/projects/pam-u2f'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/user/projects/pam-u2f'
Making all in pamu2fcfg
make[1]: Entering directory '/home/user/projects/pam-u2f/pamu2fcfg'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/user/projects/pam-u2f/pamu2fcfg'
Making all in tests
make[1]: Entering directory '/home/user/projects/pam-u2f/tests'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/user/projects/pam-u2f/tests'
Making all in man
make[1]: Entering directory '/home/user/projects/pam-u2f/man'
  GEN      pamu2fcfg.1
  GEN      pam_u2f.8
make[1]: Leaving directory '/home/user/projects/pam-u2f/man'
```

</details>


<details>
<summary><code>make install</code></summary>

```console
user@computer:~/projects/pam-u2f$ make install
Making install in .
make[1]: Entering directory '/home/user/projects/pam-u2f'
make[2]: Entering directory '/home/user/projects/pam-u2f'
 /usr/bin/mkdir -p '/lib/x86_64-linux-gnu/security'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   pam_u2f.la '/lib/x86_64-linux-gnu/security'
libtool: install: /usr/bin/install -c .libs/pam_u2f.so /lib/x86_64-linux-gnu/security/pam_u2f.so
/usr/bin/install: cannot create regular file '/lib/x86_64-linux-gnu/security/pam_u2f.so': Permission denied
make[2]: *** [Makefile:673: install-pampluginexecLTLIBRARIES] Error 1
make[2]: Leaving directory '/home/user/projects/pam-u2f'
make[1]: *** [Makefile:1091: install-am] Error 2
make[1]: Leaving directory '/home/user/projects/pam-u2f'
make: *** [Makefile:782: install-recursive] Error 1
```

</details>


<details>
<summary><code>sudo make install</code></summary>

```console
user@computer:~/projects/pam-u2f$ sudo make install
Making install in .
make[1]: Entering directory '/home/user/projects/pam-u2f'
make[2]: Entering directory '/home/user/projects/pam-u2f'
 /usr/bin/mkdir -p '/lib/x86_64-linux-gnu/security'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   pam_u2f.la '/lib/x86_64-linux-gnu/security'
libtool: install: /usr/bin/install -c .libs/pam_u2f.so /lib/x86_64-linux-gnu/security/pam_u2f.so
libtool: install: /usr/bin/install -c .libs/pam_u2f.lai /lib/x86_64-linux-gnu/security/pam_u2f.la
libtool: finish: PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/sbin" ldconfig -n /lib/x86_64-linux-gnu/security
----------------------------------------------------------------------
Libraries have been installed in:
   /lib/x86_64-linux-gnu/security

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the '-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the 'LD_RUN_PATH' environment variable
     during linking
   - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to '/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
make  install-exec-hook
make[3]: Entering directory '/home/user/projects/pam-u2f'
rm -f /lib/x86_64-linux-gnu/security/pam_u2f.la
chmod -f 644 /lib/x86_64-linux-gnu/security/pam_u2f.so || true
make[3]: Leaving directory '/home/user/projects/pam-u2f'
make[2]: Nothing to be done for 'install-data-am'.
make[2]: Leaving directory '/home/user/projects/pam-u2f'
make[1]: Leaving directory '/home/user/projects/pam-u2f'
Making install in pamu2fcfg
make[1]: Entering directory '/home/user/projects/pam-u2f/pamu2fcfg'
make[2]: Entering directory '/home/user/projects/pam-u2f/pamu2fcfg'
 /usr/bin/mkdir -p '/usr/local/bin'
  /bin/bash ../libtool   --mode=install /usr/bin/install -c pamu2fcfg '/usr/local/bin'
libtool: install: /usr/bin/install -c pamu2fcfg /usr/local/bin/pamu2fcfg
make[2]: Nothing to be done for 'install-data-am'.
make[2]: Leaving directory '/home/user/projects/pam-u2f/pamu2fcfg'
make[1]: Leaving directory '/home/user/projects/pam-u2f/pamu2fcfg'
Making install in tests
make[1]: Entering directory '/home/user/projects/pam-u2f/tests'
make[2]: Entering directory '/home/user/projects/pam-u2f/tests'
make[2]: Nothing to be done for 'install-exec-am'.
make[2]: Nothing to be done for 'install-data-am'.
make[2]: Leaving directory '/home/user/projects/pam-u2f/tests'
make[1]: Leaving directory '/home/user/projects/pam-u2f/tests'
Making install in man
make[1]: Entering directory '/home/user/projects/pam-u2f/man'
make[2]: Entering directory '/home/user/projects/pam-u2f/man'
make[2]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/usr/local/share/man/man1'
 /usr/bin/install -c -m 644 pamu2fcfg.1 '/usr/local/share/man/man1'
 /usr/bin/mkdir -p '/usr/local/share/man/man8'
 /usr/bin/install -c -m 644 pam_u2f.8 '/usr/local/share/man/man8'
make[2]: Leaving directory '/home/user/projects/pam-u2f/man'
make[1]: Leaving directory '/home/user/projects/pam-u2f/man'
```

</details>

Again, thanks so much for taking the time to implement this feature!

--- a/COPYING
+++ b/COPYING
@@ -1,4 +1,4 @@
-Copyright (c) 2014-2022 Yubico AB
+Copyright (c) 2014-2023 Yubico AB
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without

--- a/Makefile.am
+++ b/Makefile.am
@@ -23,6 +23,7 @@ libmodule_la_SOURCES = pam-u2f.c
 libmodule_la_SOURCES += b64.c b64.h
 libmodule_la_SOURCES += debug.c debug.h
 libmodule_la_SOURCES += drop_privs.h
+libmodule_la_SOURCES += expand.c
 libmodule_la_SOURCES += explicit_bzero.c
 libmodule_la_SOURCES += util.c util.h
 libmodule_la_LIBADD = -lpam $(LIBFIDO2_LIBS) $(LIBCRYPTO_LIBS)

--- a/README
+++ b/README
@@ -140,6 +140,11 @@ file>> may be configured relative to the users' home dirs, e.g.
 $XDG_CONFIG_HOME/Yubico/u2f_keys. If $XDG_CONFIG_HOME is not set,
 $HOME/.config/Yubico/u2f_keys is used.
 
+expand::
+Enables variable expansion within the authfile path: `%u` is expanded to the
+local user name (`PAM_USER`) and `%%` is expanded to `%`. Unknown expansion
+sequences result in an authentication error. See also `openasuser`.
+
 authpending_file=file::
 Set the location of the file that is used for touch request
 notifications. This file will be opened when pam-u2f starts waiting

--- a/README
+++ b/README
@@ -167,8 +167,8 @@ Setuid to the authenticating user when opening the authfile. Useful
 when the user's home is stored on an NFS volume mounted with the
 `root_squash` option (which maps root to nobody which will not be able
 to read the file). Note that after release 1.0.8 this is done by
-default when no global authfile or XDG_CONFIG_HOME environment
-variable has been set.
+default when no global authfile (path is absolute) or XDG_CONFIG_HOME
+environment variable has been set.
 
 alwaysok::
 Set to enable all authentication attempts to succeed (aka presentation

--- a/expand.c
+++ b/expand.c
@@ -1,0 +1,70 @@
+/*
+ * Copyright (C) 2023 Yubico AB - See COPYING
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+
+#include "util.h"
+
+static int buf_write(uint8_t **dst, size_t *size, const void *src, size_t n) {
+  if (*size < n) {
+    return -1;
+  }
+
+  memcpy(*dst, src, n);
+  *dst += n;
+  *size -= n;
+
+  return 0;
+}
+
+static int buf_write_byte(uint8_t **dst, size_t *size, uint8_t c) {
+  return buf_write(dst, size, &c, sizeof(c));
+}
+
+static const char *lookup(char var, const char *user) {
+  switch (var) {
+    case 'u':
+      return user;
+    case '%':
+      return "%";
+    default:
+      // Capture all unknown variables (incl. null byte).
+      return NULL;
+  }
+}
+
+char *expand_variables(const char *str, const char *user) {
+  uint8_t *tail, *head;
+  size_t size = PATH_MAX;
+  int ok = -1;
+
+  if (str == NULL || (tail = head = malloc(size)) == NULL) {
+    return NULL;
+  }
+
+  for (; *str != '\0'; str++) {
+    if (*str == '%') {
+      str++;
+      const char *value = lookup(*str, user);
+      if (value == NULL || *value == '\0' ||
+          buf_write(&head, &size, value, strlen(value)) != 0) {
+        goto fail;
+      }
+    } else if (buf_write_byte(&head, &size, *str) != 0) {
+      goto fail;
+    }
+  }
+
+  ok = buf_write_byte(&head, &size, '\0');
+
+fail:
+  if (ok != 0) {
+    free(tail);
+    return NULL;
+  }
+  return (char *) tail;
+}

--- a/man/pam_u2f.8.txt
+++ b/man/pam_u2f.8.txt
@@ -48,6 +48,11 @@ $XDG_CONFIG_HOME/Yubico/u2f_keys. If $XDG_CONFIG_HOME is not set,
 $HOME/.config/Yubico/u2f_keys is used. The authfile format is
 <username>:<KeyHandle1>,<UserKey1>,<CoseType1>,<Options1>:<KeyHandle2>,<UserKey2>,<CoseType2>,<Options2>:...
 
+*expand*::
+Enables variable expansion within the authfile path: `%u` is expanded to the
+local user name (`PAM_USER`) and `%%` is expanded to `%`. Unknown expansion
+sequences result in an authentication error. See also `openasuser`.
+
 *authpending_file*=_file_::
 Set the location of the file that is used for touch request
 notifications. This file will be opened when pam-u2f starts waiting

--- a/man/pam_u2f.8.txt
+++ b/man/pam_u2f.8.txt
@@ -72,8 +72,8 @@ Setuid to the authenticating user when opening the authfile. Useful
 when the user's home is stored on an NFS volume mounted with the
 root_squash option (which maps root to nobody which will not be able
 to read the file). Note that after release 1.0.8 this is done by
-default when no global authfile or XDG_CONFIG_HOME environment
-variable has been set.
+default when no global authfile (path is absolute) or XDG_CONFIG_HOME
+environment variable has been set.
 
 *alwaysok*::
 Set to enable all authentication attempts to succeed (aka presentation

--- a/pam-u2f.c
+++ b/pam-u2f.c
@@ -1,5 +1,5 @@
 /*
- *  Copyright (C) 2014-2021 Yubico AB - See COPYING
+ *  Copyright (C) 2014-2023 Yubico AB - See COPYING
  */
 
 /* Define which PAM interfaces we provide */
@@ -23,6 +23,8 @@
 #include "debug.h"
 #include "drop_privs.h"
 #include "util.h"
+
+#define free_const(a) free((void *) (uintptr_t) (a))
 
 /* If secure_getenv is not defined, define it here */
 #ifndef HAVE_SECURE_GETENV
@@ -61,6 +63,8 @@ static void parse_cfg(int flags, int argc, const char **argv, cfg_t *cfg) {
       cfg->cue = 1;
     } else if (strcmp(argv[i], "nodetect") == 0) {
       cfg->nodetect = 1;
+    } else if (strcmp(argv[i], "expand") == 0) {
+      cfg->expand = 1;
     } else if (strncmp(argv[i], "userpresence=", 13) == 0) {
       sscanf(argv[i], "userpresence=%d", &cfg->userpresence);
     } else if (strncmp(argv[i], "userverification=", 17) == 0) {
@@ -106,6 +110,7 @@ static void parse_cfg(int flags, int argc, const char **argv, cfg_t *cfg) {
   debug_dbg(cfg, "openasuser=%d", cfg->openasuser);
   debug_dbg(cfg, "alwaysok=%d", cfg->alwaysok);
   debug_dbg(cfg, "sshformat=%d", cfg->sshformat);
+  debug_dbg(cfg, "expand=%d", cfg->expand);
   debug_dbg(cfg, "authfile=%s", cfg->auth_file ? cfg->auth_file : "(null)");
   debug_dbg(cfg, "authpending_file=%s",
             cfg->authpending_file ? cfg->authpending_file : "(null)");
@@ -256,13 +261,27 @@ int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
   debug_dbg(cfg, "Found user %s", user);
   debug_dbg(cfg, "Home directory for %s is %s", user, pw->pw_dir);
 
+  // Perform variable expansion.
+  if (cfg->expand && cfg->auth_file) {
+    if ((cfg->auth_file = expand_variables(cfg->auth_file, user)) == NULL) {
+      debug_dbg(cfg, "Failed to perform variable expansion");
+      retval = PAM_AUTHINFO_UNAVAIL;
+      goto done;
+    }
+    should_free_auth_file = 1;
+  }
+  // Resolve default or relative paths.
   if (!cfg->auth_file || cfg->auth_file[0] != '/') {
-    if ((cfg->auth_file = resolve_authfile_path(cfg, pw, &openasuser)) ==
-        NULL) {
+    char *tmp = resolve_authfile_path(cfg, pw, &openasuser);
+    if (tmp == NULL) {
       debug_dbg(cfg, "Could not resolve authfile path");
       retval = PAM_IGNORE;
       goto done;
     }
+    if (should_free_auth_file) {
+      free_const(cfg->auth_file);
+    }
+    cfg->auth_file = tmp;
     should_free_auth_file = 1;
   }
 
@@ -380,7 +399,6 @@ int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
 done:
   free_devices(devices, n_devices);
 
-#define free_const(a) free((void *) (uintptr_t)(a))
   if (should_free_origin) {
     free_const(cfg->origin);
     cfg->origin = NULL;

--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -13,4 +13,7 @@ dlsym_check_LDFLAGS = -ldl $(AM_LDFLAGS)
 check_PROGRAMS += get_devices
 get_devices_LDADD = $(top_builddir)/libmodule.la
 
+check_PROGRAMS += expand
+expand_LDADD = $(top_builddir)/libmodule.la
+
 TESTS = $(check_PROGRAMS)

--- a/tests/expand.c
+++ b/tests/expand.c
@@ -1,0 +1,42 @@
+/*
+ *  Copyright (C) 2023 Yubico AB - See COPYING
+ */
+
+#undef NDEBUG
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include "util.h"
+
+#define ASSERT_STR_EQ(a, b) assert(!strcmp(a, b))
+#define ASSERT_EXPANDED_EQ(str, user, result)                                  \
+  do {                                                                         \
+    char *tmp = expand_variables(str, user);                                   \
+    assert(tmp != NULL);                                                       \
+    ASSERT_STR_EQ(tmp, result);                                                \
+    free(tmp);                                                                 \
+  } while (0)
+
+#define ASSERT_NULL(x) assert((x) == NULL)
+
+int main(void) {
+  ASSERT_EXPANDED_EQ("foobar", "user", "foobar");
+  ASSERT_EXPANDED_EQ("", "user", "");
+  ASSERT_EXPANDED_EQ("%%", "user", "%");
+  ASSERT_EXPANDED_EQ("%u", "user", "user");
+  ASSERT_EXPANDED_EQ("x%u", "user", "xuser");
+  ASSERT_EXPANDED_EQ("%ux", "user", "userx");
+  ASSERT_EXPANDED_EQ("x%ux", "user", "xuserx");
+  ASSERT_EXPANDED_EQ("%%%u", "user", "%user");
+  ASSERT_EXPANDED_EQ("%u%%", "user", "user%");
+  ASSERT_EXPANDED_EQ("%%u", "user", "%u");
+  ASSERT_EXPANDED_EQ("%u", "%user", "%user");
+  ASSERT_EXPANDED_EQ("%u%u", "user", "useruser");
+  ASSERT_EXPANDED_EQ("%%%u%%", "user", "%user%");
+
+  ASSERT_NULL(expand_variables("%", "user"));  // Unexpected end of string.
+  ASSERT_NULL(expand_variables("%x", "user")); // Unknown variable.
+  ASSERT_NULL(expand_variables("%u", ""));     // Disallow empty username.
+
+  return 0;
+}

--- a/util.h
+++ b/util.h
@@ -35,6 +35,7 @@ typedef struct {
   int userverification;
   int pinverification;
   int sshformat;
+  int expand;
   const char *auth_file;
   const char *authpending_file;
   const char *origin;

--- a/util.h
+++ b/util.h
@@ -64,6 +64,7 @@ char *converse(pam_handle_t *pamh, int echocode, const char *prompt);
 int random_bytes(void *, size_t);
 int cose_type(const char *, int *);
 const char *cose_string(int);
+char *expand_variables(const char *, const char *);
 
 #if !defined(HAVE_EXPLICIT_BZERO)
 void explicit_bzero(void *, size_t);

